### PR TITLE
feat: Multiple RequestMirrors Filters per HTTPRoute Rule

### LINKED ISSUE #1811: Support Multiple RequestMirror filters per HTTPRoute Rule
*Description*:

* Allow the user to specify multiple https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.HTTPRequestMirrorFilter within a single rule

Changes needed
* Create a Mirror Backend/ IR RouteDestination per Filter https://github.com/envoyproxy/gateway/blob/4285f4e269ff309e81ec24b2f80fa9b5343f6a8f/internal/gatewayapi/filters.go#L810, possibly using `filterIdx` as a unique identifier
* Make the Mirror field a list in https://github.com/envoyproxy/gateway/blob/4285f4e269ff309e81ec24b2f80fa9b5343f6a8f/internal/ir/xds.go#L264

* Build a new xds mirror policy per mirror route destination https://github.com/envoyproxy/gateway/blob/4285f4e269ff309e81ec24b2f80fa9b5343f6a8f/internal/xds/translator/route.go#L296

### PR DESCRIPTION
**Pull request by Ronnie-personal**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #1811

### CODE REVIEW

**Comment by Ronnie-personal** [comment]
@arkodg 
Could you please take a quick look at this initial draft code change? Am I heading in the right direction?
I will spend more time on the testing.

File: internal/gatewayapi/filters.go [hunk 1]
```diff
@@ -76,7 +76,7 @@ func (t *Translator) ProcessHTTPFilters(parentRef *RouteParentContext,
 		RuleIdx:      ruleIdx,
 		HTTPFilterIR: &HTTPFilterIR{},
 	}
-
+	filterIdx := 0
 	for i := range filters {
```
Comment (arkodg) [inline internal/gatewayapi/filters.go:80]: why not just use i

File: internal/gatewayapi/filters.go [hunk 2]
```diff
@@ -839,22 +842,26 @@ func (t *Translator) processRequestMirrorFilter(
 	// Only add missing mirror destinations
```
Comment (arkodg) [inline internal/gatewayapi/filters.go:842]: imo this entire logic of adding missing mirror destinations can be removed, and so you can simplify this logic

File: internal/ir/xds.go [hunk 3]
```diff
@@ -261,7 +261,7 @@ type HTTPRoute struct {
 	// Redirections to be returned for this route. Takes precedence over Destinations.
 	Redirect *Redirect `json:"redirect,omitempty" yaml:"redirect,omitempty"`
 	// Destination that requests to this HTTPRoute will be mirrored to
-	Mirror *RouteDestination `json:"mirror,omitempty" yaml:"mirror,omitempty"`
+	Mirror []*RouteDestination `json:"mirror,omitempty" yaml:"mirror,omitempty"`
```
Comment (arkodg) [inline internal/ir/xds.go:264]:
```suggestion
	Mirrors []*RouteDestination `json:"mirrors,omitempty" yaml:"mirrors,omitempty"`
```

**Comment by arkodg** [comment]
looks good @Ronnie-personal , added some comments, thanks for picking this one up

File: internal/gatewayapi/filters.go (hunk 2 above)
Comment (Ronnie-personal) [reply internal/gatewayapi/filters.go:842]: > imo this entire logic of adding missing mirror destinations can be removed, and so you can simplify this logic

Are we building filterContext.Mirror from scratch and there is no duplicate Endpoints?
I definitely can refactor the code based on your assessment.

File: internal/gatewayapi/filters.go (hunk 1 above)
Comment (Ronnie-personal) [reply internal/gatewayapi/filters.go:80]: > why not just use i

Yes, I can use 'i' if I don't need to match the value with the index of Mirror list.

File: internal/ir/xds.go (hunk 3 above)
Comment (Ronnie-personal) [reply internal/ir/xds.go:264]: Good catch, I will fix the code.

File: internal/gatewayapi/filters.go (hunk 2 above)
Comment (arkodg) [reply internal/gatewayapi/filters.go:842]: yes we dont need to consider the duplicate case, please go ahead and refactor

**Comment by Ronnie-personal** [comment]
Can anyone help me to understand this pipeline error at https://github.com/envoyproxy/gateway/actions/runs/5971186116/job/16224598674?pr=1819#step:6:140?

I tested 'make -k lint' locally, I don't see the error.

**Comment by arkodg** [comment]
here is the error
```
./internal/xds/translator/testdata/in/xds-ir/http-route-mirror.yaml
[127](https://github.com/envoyproxy/gateway/actions/runs/5971186116/job/16224598674?pr=1819#step:6:128)
  Error: 17:7 [indentation] wrong indentation: expected 4 but found 6
```

**Comment by Ronnie-personal** [comment]
> here is the error
> 
> ```
> ./internal/xds/translator/testdata/in/xds-ir/http-route-mirror.yaml
> [127](https://github.com/envoyproxy/gateway/actions/runs/5971186116/job/16224598674?pr=1819#step:6:128)
>   Error: 17:7 [indentation] wrong indentation: expected 4 but found 6
> ```

Thank you! I will fix it.

**Comment by Ronnie-personal** [comment]
What's happening with the pipeline? it only executed one check.

File: internal/ir/xds.go [hunk 4]
```diff
@@ -261,7 +261,7 @@ type HTTPRoute struct {
 	// Redirections to be returned for this route. Takes precedence over Destinations.
 	Redirect *Redirect `json:"redirect,omitempty" yaml:"redirect,omitempty"`
 	// Destination that requests to this HTTPRoute will be mirrored to
-	Mirror *RouteDestination `json:"mirror,omitempty" yaml:"mirror,omitempty"`
+	Mirror []*RouteDestination `json:"mirrors,omitempty" yaml:"mirrors,omitempty"`
```
Comment (arkodg) [inline internal/ir/xds.go:264]:
```suggestion
	Mirrors []*RouteDestination `json:"mirrors,omitempty" yaml:"mirrors,omitempty"`
```

File: internal/xds/translator/testdata/in/xds-ir/http-route-mirror.yaml [hunk 5]
```diff
@@ -13,7 +13,10 @@ http:
       endpoints:
       - host: "1.2.3.4"
         port: 50000
-    mirror:
-      name: "mirror-route-dest"
+    mirrors:
```
Comment (arkodg) [inline internal/xds/translator/testdata/in/xds-ir/http-route-mirror.yaml:16]: lets keep this example as is, and create one more test file for multiple-mirrors ?

File: internal/gatewayapi/filters.go [hunk 6]
```diff
@@ -836,27 +835,13 @@ func (t *Translator) processRequestMirrorFilter(
...
-					Name: fmt.Sprintf("%s-mirror", irRouteDestinationName(filterContext.Route, filterContext.RuleIdx)),
-				}
-			}
-			filterContext.Mirror.Endpoints = append(filterContext.Mirror.Endpoints, mirrorEp)
-		}
+	newMirror := &ir.RouteDestination{
+		Name: fmt.Sprintf("%s-mirror-%d", irRouteDestinationName(filterContext.Route, filterContext.RuleIdx), filterIdx),
 	}
+	filterContext.Mirror = append(filterContext.Mirror, newMirror)
+	// Get the index of the last mirror added
```
Comment (arkodg) [inline internal/gatewayapi/filters.go:842]: why do we need L842-L844, can't we just assign `Endpoints` on L839 when creating `RouteDestination` ?

File: internal/xds/translator/testdata/in/xds-ir/http-route-mirror.yaml (hunk 5 above)
Comment (Ronnie-personal) [reply internal/xds/translator/testdata/in/xds-ir/http-route-mirror.yaml:16]: Sound good, will do.

**Comment by zirain** [comment]
@Ronnie-personal can you remove following line, let's check with CI?

**Comment by Ronnie-personal** [comment]
> @Ronnie-personal can you remove following line, let's check with CI?

Sorry, which line do you want to remove?

**Comment by zirain** [comment]
should be 
https://github.com/envoyproxy/gateway/blob/1f08e46698bffa44484d5f90f89bac8c9ebe12a4/test/conformance/conformance_test.go#L49-L50

**Comment by Ronnie-personal** [comment]
> should be
> 
> https://github.com/envoyproxy/gateway/blob/1f08e46698bffa44484d5f90f89bac8c9ebe12a4/test/conformance/conformance_test.go#L49-L50

Thanks, I'm working on it right now.

Do we also need to remove https://github.com/envoyproxy/gateway/blob/main/test/conformance/experimental_conformance_test.go#L96

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by zirain — APPROVED** [review]

**Pull request merged**